orttraining/orttraining/python/training/onnxblock/_graph_utils.py

```diff
@@ -0,0 +1,144 @@
+
+    # Collect names of parameters that need gradients computed
+    all_args_requiring_gradient = set()
+    # Move all trainable and non trainable initializers to graph inputs.
+    # This allows training to pass in the parameters from outside the graph
+    # so as to share the parameters across multiple sessions.
+    graph_inputs = model.graph.input
+    initializers = []
+    for initializer in model.graph.initializer:
+        if not initializer.name[0].isdigit():
+            # Move only those initializers as inputs that are not local
+            # to the onnx model. i.e. initializers that are model parameters.
+            # These are tpically those initializers without any number prefixed
+            # to their names.
+            graph_inputs.append(
+                onnx.helper.make_tensor_value_info(
+                    initializer.name, initializer.data_type, initializer.dims
+                )
+            )
+            if initializer.name not in user_args_not_requiring_grad:
+                all_args_requiring_gradient.add(initializer.name)
+        else:
+            # All other initializers stay where they were.
+            initializers.append(initializer)
+
+    # Graph and model with initializers as inputs.
+    graph_with_initializers_as_inputs = onnx.helper.make_graph(
+        model.graph.node,
+        "Forward Graph with Initilializers as Inputs",
+        graph_inputs,
+        model.graph.output,
+        initializer=initializers,
+    )
+    grad_model = onnx.helper.make_model(
+        graph_with_initializers_as_inputs,
+        producer_name=onnxblock._producer_name,
+        opset_imports=[onnx.helper.make_opsetid("com.microsoft", 1)]
+        + list(model.opset_import),
+    )
+
+    # Any graph input that requires gradient, should have been already added to
+    # args_requiring_grad.
+    for argument_name in user_args_requiring_grad:
+        all_args_requiring_gradient.add(argument_name)
+
+    # Assumption is that the graph has an output called `loss`.
+    builder = GradientGraphBuilder(
+        grad_model.SerializeToString(), {"loss"}, all_args_requiring_gradient, "loss"
+    )
+    builder.build()
+    return onnx.load_from_string(builder.get_model())
+
+
+def build_gradient_accumulation_model(grad_model):
+    """Builds gradient accumulation nodes on top of a training model."""
+
+    graph_inputs = grad_model.graph.input
+    graph_nodes = grad_model.graph.node
+    graph_outputs = []
+
+    lazy_reset_grad_input_name = "lazy_reset_grad"
+    gradient_output_suffix = "_grad"
+    gradient_accumulation_name = "accumulation"
+    gradient_buffer_name_suffix = "buffer"
+    gradient_output_name_suffix = "out"
+
+    for idx, graph_output in enumerate(grad_model.graph.output):
+        # if the graph output ends with _grad,
+        # assume that that output is a gradient output
+        if not graph_output.name.endswith(gradient_output_suffix):
```

Comment:
Maybe we can just pass the _arg_requiring_grad in, to help double validate the graph_output.name? "_grad" postfix is fine, while I feel the check is a bit weak.

Reply:
Sorry I did not understand why "_args_requiring_grad+model parameters-_args_not_requiring_grad should represent the entire list of arguments that have their gradients computed.", shouldn't _args_requiring_grad returned all trainable weights having gradients?

Reply:
`_args_requiring_grad` and `_args_not_requiring_grad` are based on the user's setting of requires_grad for a particular argument.

While trying to build gradient, I try to auto detect the model parameters without needing the user intervention (without needing which arguments require gradient from the user). So, the `_args_requiring_grad` does not contain all the arguments that require gradient. It only contains those that the user set.

Fortunately, before building the gradient, I pull all therequired gradients. So, I can just reuse those. Made this change in the new commit.